### PR TITLE
Re-allow builds with Anaconda Python (Only Python 3 checked)

### PR DESCRIPTION
**Pull request by fcr**
After this change I  successfully built the head branch using Anaconda on my Mac.

Note you may want to add to the build instructions that the following be run before the build on Mac:
export ARCHFLAGS="-arch x86_64"
export MACOSX_DEPLOYMENT_TARGET=10.14

In addition I "solved" the problem of the SpacialPooler.getPermanence() always returning zero values.
It turns out that PyBind11 will convert an array passed as an argument if the data types do not exactly match up with the C++ definitions. Then the C++ function only gets a converted copy of the numpy array, and any changes are lost after returning. This was what was happening to me.

If you pass the numpy array with the type matching the C++ argument then PyBind11 does not convert the data, and the C++ code can modify the data in-place. Once I discovered this, my ocr program gave exactly the same results in Python 3.77 as the original nupic Python 2.7 version.

I found this solution in this stack overflow query: https://stackoverflow.com/questions/54793539/pybind11-modify-numpy-array-from-c.

To test for this and to provide examples for others, I added a few unit tests to the Python Spacial Pooler bindings.

### CODE REVIEW

**Comment by breznak** [comment]
> Once I discovered this, my ocr program gave exactly the same results in Python 3.77 as the original nupic Python 2.7 version.

I'm really happy to have another use of this repo with existing python code! :clap: 
PS: If you would be interested, we're keen to have some vision encoders in the repo (#259 ) or you can have a look at the src/examples/mnist/ which is similar (same?) to the OCR code from nupic.vision?

**Comment by fcr** [comment]
Okay I have managed to find how to turn off the automatic PyBind11 conversions.
Just use py:array instead of py:array_t. The py:array_t will always try to convert the array to type t.
I added definitions to Types.hpp so that I can know the correct data size and check for it before passing the array to the function.
I put the checking code into the getit() utility function.
I modified the python unit tests to check for the data tpye being correctly checked.
I did not like adding the definitions to Types.hpp but this was the best I could think off with my limited C++ experience.

File: src/nupic/types/Types.hpp [hunk 1]
```diff
@@ -123,8 +123,10 @@ typedef std::size_t Size;
  */
 #ifdef NTA_DOUBLE_PRECISION
   typedef Real64 Real;
+  const int FLOAT_PRECISION = 8; // bytes
```
Comment (ctrl-z-9000-times) [inline src/nupic/types/Types.hpp]: nitpick: try using the `sizeof` macro: `sizeof( Real ) -> number of bytes`

Comment (fcr) [reply src/nupic/types/Types.hpp]: I agree. Didn't like this solution either. Thanks for the solution.
I'll your solution to use.

Comment (fcr) [reply src/nupic/types/Types.hpp]: Okay. I have restored the original Types.hpp and used sizeof in py_SpacialPooler.cpp

**Comment by breznak** [comment]
@fcr I somehow cannot make a PR to your repo.. 
Could you please try to cleanup this? Probably this would do the job
```
(on master) git pull <from community> 
(on this branch) git rebabase -i community/master
git push --force 
```

Alternatively there's a clean branch for review https://github.com/htm-community/nupic.cpp/tree/pybind_type_fixes

**Comment by fcr** [comment]
I rebased before merging but it I guess I was too slow. :-(

**Comment by fcr** [comment]
I ran the cleanup but am waiting for the build to finish to make sure all is okay before pushing.
Meanwhile I reviewed the changes in https://github.com/htm-community/nupic.cpp/tree/pybind_type_fixes and they look correct.

**Comment by fcr** [comment]
I have run git push --force.
See if this helps.

**Comment by Sedom9** [comment]
Hi! Did you test Anaconda on Windows ?

**Comment by fcr** [comment]
No. I only use MacOS.

**Pull request merged**